File: packages/gavio-py/gavio/inspector/buffer.py
"""RingBuffer — bounded in-memory trace store fed by the InspectorBus."""

from __future__ import annotations

import threading
from collections import OrderedDict
from typing import Any

MAX_EVENTS_PER_TRACE = 500
COST_DIMENSION_KEYS = ("feature", "tenant", "user", "endpoint", "environment", "workflow", "tool")

_SUMMARY_START_KEYS = (
    ("parentTraceId", None),
    ("agentId", None),
    ("sessionId", None),
    ("provider", None),
    ("model", None),
    ("wallTimeUtc", None),
)
_SUMMARY_END_KEYS = (
    "status",
    "latencyMs",
    "costUsd",
    "cacheHit",
    "cacheType",
    "piiEntityTypes",
    "interceptorsFired",
)


class RingBuffer:
    """Assembles inspector events into per-trace records, oldest evicted first.

    Each trace is ``{"summary": {...}, "events": [...]}``. The summary is built
    from trace.start and trace.end data; events are capped at
    :data:`MAX_EVENTS_PER_TRACE` per trace. Thread-safe — the HTTP server reads
    from other threads.
    """

    def __init__(self, max_traces: int = 1000) -> None:
        self._max_traces = max_traces
        self._traces: OrderedDict[str, dict[str, Any]] = OrderedDict()
        self._lock = threading.Lock()

    def on_event(self, event: dict[str, Any]) -> None:
        """Bus subscriber — append the event to its trace."""
        trace_id = event["traceId"]
        with self._lock:
            trace = self._traces.get(trace_id)
            if trace is None:
                while len(self._traces) >= self._max_traces:
                    self._traces.popitem(last=False)
                trace = {
                    "summary": _with_cost_defaults({"traceId": trace_id, "status": "running"}),
                    "events": [],
                }
                self._traces[trace_id] = trace
            if len(trace["events"]) < MAX_EVENTS_PER_TRACE:
                trace["events"].append(event)
            self._update_summary(trace["summary"], event)

    @staticmethod
    def _update_summary(summary: dict[str, Any], event: dict[str, Any]) -> None:
        data = event["data"]
        if event["type"] == "trace.start":
            for key, default in _SUMMARY_START_KEYS:
                summary[key] = data.get(key, default)
            dimensions = _cost_dimensions(data.get("costDimensions"))
            summary["costDimensions"] = dimensions
            for key in COST_DIMENSION_KEYS:
                summary[key] = dimensions.get(key)
        elif event["type"] == "trace.end":
            for key in _SUMMARY_END_KEYS:
                if key in data:
                    summary[key] = data[key]
            fired = summary.get("interceptorsFired") or []
            summary["middlewareChain"] = ">".join(fired) if fired else None
            if "cacheSavingsUsd" in data:
                summary["cacheSavingsUsd"] = data["cacheSavingsUsd"]
        elif event["type"] == "provider.call.start":
            attempt = _number(data.get("attempt")) or (summary.get("providerCallCount", 0) + 1)
            summary["providerCallCount"] = max(summary.get("providerCallCount", 0), attempt)
            summary["retryCount"] = max(0, summary["providerCallCount"] - 1)
        elif event["type"] == "provider.call.end":
            if "usage" in data:
                # Token usage feeds /api/stats and /api/simulate-cost.
                summary["usage"] = data["usage"]
            attempt = _number(data.get("attempt"))
            cost_usd = _number(data.get("costUsd")) or 0
            if attempt is not None and attempt > 1 and cost_usd > 0:
                summary["retryOverheadUsd"] = round(
                    summary.get("retryOverheadUsd", 0.0) + cost_usd,
                    8,
                )
        elif event["type"] == "governance.event" and data.get("kind") == "drift":
            metric = data.get("metric")
            if isinstance(metric, str):
                summary.setdefault("driftAlerts", []).append(metric)

    def seed(self, summary: dict[str, Any], events: list[dict[str, Any]] | None = None) -> None:
        """Insert a pre-assembled trace — store mode loads audit records this way."""
        with self._lock:
            while len(self._traces) >= self._max_traces:
                self._traces.popitem(last=False)
            self._traces[summary["traceId"]] = {
                "summary": _with_cost_defaults(dict(summary)),
                "events": list(events or []),
            }

    def count(self) -> int:
        with self._lock:
            return len(self._traces)

    def summaries(self, limit: int | None = None) -> list[dict[str, Any]]:
        """Trace summaries, chronological ascending. ``limit`` keeps the most recent N."""
        with self._lock:
            out = [dict(t["summary"]) for t in self._traces.values()]
        if limit is not None and limit >= 0:
            out = out[-limit:] if limit else []
        return out

    def get(self, trace_id: str) -> dict[str, Any] | None:
        """Full record for one trace, or None if unknown/evicted."""
        with self._lock:
            trace = self._traces.get(trace_id)
            if trace is None:
                return None
            return {"summary": dict(trace["summary"]), "events": list(trace["events"])}


def _with_cost_defaults(summary: dict[str, Any]) -> dict[str, Any]:
    summary.setdefault("costDimensions", {})
    for key in COST_DIMENSION_KEYS:
        summary.setdefault(key, (summary.get("costDimensions") or {}).get(key))
    summary.setdefault("middlewareChain", None)
    summary.setdefault("providerCallCount", 0)
    summary.setdefault("retryCount", 0)
    summary.setdefault("retryOverheadUsd", 0.0)
    summary.setdefault("cacheSavingsUsd", 0.0)
    return summary


def _cost_dimensions(value: Any) -> dict[str, str]:
    if not isinstance(value, dict):
        return {}
    return {key: value[key] for key in COST_DIMENSION_KEYS if isinstance(value.get(key), str)}


def _number(value: Any) -> int | float | None:
    return value if isinstance(value, (int, float)) else None

File: packages/gavio-js/src/inspector/buffer.ts
/**
 * TraceBuffer — a bounded ring buffer that subscribes to the InspectorBus and
 * assembles events into per-trace records: { summary, events[] }.
 *
 * Bounded on both axes: at most `maxTraces` traces (oldest evicted first) and
 * at most `maxEventsPerTrace` events kept per trace.
 */

import type { InspectorEvent } from './events.js'
import { COST_DIMENSION_KEYS, type CostDimensions } from './events.js'

/** Roll-up of one trace, fed by trace.start and trace.end events. */
export interface TraceSummary {
  traceId: string
  parentTraceId: string | null
  agentId: string | null
  sessionId: string | null
  provider: string | null
  model: string | null
  status: string
  latencyMs: number | null
  costUsd: number | null
  cacheHit: boolean | null
  cacheType: string | null
  piiEntityTypes: string[]
  wallTimeUtc: string | null
  interceptorsFired: string[]
  costDimensions: CostDimensions
  feature: string | null
  tenant: string | null
  user: string | null
  endpoint: string | null
  environment: string | null
  workflow: string | null
  tool: string | null
  middlewareChain: string | null
  providerCallCount: number
  retryCount: number
  retryOverheadUsd: number
  cacheSavingsUsd: number
  /** Token usage from provider.call.end — feeds /api/stats and /api/simulate-cost. */
  usage?: { promptTokens: number; completionTokens: number; totalTokens: number }
  /** Content hashes (audit-store summaries) — searchable via /api/traces?q=. */
  promptHash?: string
  responseHash?: string
  /** Metrics that drifted on this trace (F-GOV-07), from governance.event events. */
  driftAlerts?: string[]
}

export interface TraceRecord {
  summary: TraceSummary
  events: InspectorEvent[]
}

export interface TraceBufferOptions {
  maxTraces?: number
  maxEventsPerTrace?: number
}

export const DEFAULT_MAX_EVENTS_PER_TRACE = 500

export class TraceBuffer {
  private readonly maxTraces: number
  private readonly maxEventsPerTrace: number
  /** Insertion-ordered (Map preserves insertion order) — oldest trace first. */
  private readonly traces = new Map<string, TraceRecord>()

  constructor(options: TraceBufferOptions = {}) {
    this.maxTraces = options.maxTraces ?? 1000
    this.maxEventsPerTrace = options.maxEventsPerTrace ?? DEFAULT_MAX_EVENTS_PER_TRACE
  }

  /** Bus subscriber — bind or wrap with an arrow when subscribing. */
  handle(event: InspectorEvent): void {
    const record = this.recordFor(event.traceId)
    if (record.events.length < this.maxEventsPerTrace) {
      record.events.push(event)
    }
    if (event.type === 'trace.start') this.applyStart(record.summary, event.data)
    if (event.type === 'trace.end') this.applyEnd(record.summary, event.data)
    if (event.type === 'provider.call.start') this.applyProviderCallStart(record.summary, event.data)
    if (event.type === 'provider.call.end' && event.data['usage'] !== undefined) {
      record.summary.usage = event.data['usage'] as TraceSummary['usage']
    }
    if (event.type === 'provider.call.end') {
      this.applyProviderCallEnd(record.summary, event.data)
    }
    if (event.type === 'governance.event' && event.data['kind'] === 'drift') {
      const metric = event.data['metric']
      if (typeof metric === 'string') {
        ;(record.summary.driftAlerts ??= []).push(metric)
      }
    }
  }

  /** Trace summaries in chronological (ascending) order; limit keeps the most recent N. */
  list(limit?: number): TraceSummary[] {
    const all = [...this.traces.values()].map((r) => r.summary)
    if (limit !== undefined && limit >= 0 && limit < all.length) {
      return all.slice(all.length - limit)
    }
    return all
  }

  get(traceId: string): TraceRecord | undefined {
    return this.traces.get(traceId)
  }

  get size(): number {
    return this.traces.size
  }

  private recordFor(traceId: string): TraceRecord {
    let record = this.traces.get(traceId)
    if (record !== undefined) return record
    record = { summary: emptySummary(traceId), events: [] }
    this.traces.set(traceId, record)
    while (this.traces.size > this.maxTraces) {
      const oldest = this.traces.keys().next().value as string
      this.traces.delete(oldest)
    }
    return record
  }

  private applyStart(summary: TraceSummary, data: Record<string, unknown>): void {
    summary.parentTraceId = (data['parentTraceId'] as string | null | undefined) ?? null
    summary.agentId = (data['agentId'] as string | null | undefined) ?? null
    summary.sessionId = (data['sessionId'] as string | null | undefined) ?? null
    summary.provider = (data['provider'] as string | undefined) ?? null
    summary.model = (data['model'] as string | undefined) ?? null
    summary.wallTimeUtc = (data['wallTimeUtc'] as string | undefined) ?? null
    summary.costDimensions = normalizeCostDimensions(data['costDimensions'])
    for (const key of COST_DIMENSION_KEYS) {
      ;(summary as unknown as Record<typeof key, string | null>)[key] =
        summary.costDimensions[key] ?? null
    }
  }

  private applyEnd(summary: TraceSummary, data: Record<string, unknown>): void {
    summary.status = (data['status'] as string | undefined) ?? summary.status
    summary.latencyMs = (data['latencyMs'] as number | undefined) ?? null
    summary.costUsd = (data['costUsd'] as number | undefined) ?? null
    summary.cacheHit = (data['cacheHit'] as boolean | undefined) ?? null
    summary.cacheType = (data['cacheType'] as string | null | undefined) ?? null
    summary.piiEntityTypes = (data['piiEntityTypes'] as string[] | undefined) ?? []
    summary.interceptorsFired = (data['interceptorsFired'] as string[] | undefined) ?? []
    summary.middlewareChain =
      summary.interceptorsFired.length > 0 ? summary.interceptorsFired.join('>') : null
    summary.cacheSavingsUsd = numberValue(data['cacheSavingsUsd']) ?? summary.cacheSavingsUsd
  }

  private applyProviderCallStart(summary: TraceSummary, data: Record<string, unknown>): void {
    const attempt = numberValue(data['attempt']) ?? summary.providerCallCount + 1
    summary.providerCallCount = Math.max(summary.providerCallCount, attempt)
    summary.retryCount = Math.max(0, summary.providerCallCount - 1)
  }

  private applyProviderCallEnd(summary: TraceSummary, data: Record<string, unknown>): void {
    const attempt = numberValue(data['attempt'])
    const costUsd = numberValue(data['costUsd']) ?? 0
    if (attempt !== undefined && attempt > 1 && costUsd > 0) {
      summary.retryOverheadUsd = round8(summary.retryOverheadUsd + costUsd)
    }
  }
}

function emptySummary(traceId: string): TraceSummary {
  return {
    traceId,
    parentTraceId: null,
    agentId: null,
    sessionId: null,
    provider: null,
    model: null,
    status: 'running',
    latencyMs: null,
    costUsd: null,
    cacheHit: null,
    cacheType: null,
    piiEntityTypes: [],
    wallTimeUtc: null,
    interceptorsFired: [],
    costDimensions: {},
    feature: null,
    tenant: null,
    user: null,
    endpoint: null,
    environment: null,
    workflow: null,
    tool: null,
    middlewareChain: null,
    providerCallCount: 0,
    retryCount: 0,
    retryOverheadUsd: 0,
    cacheSavingsUsd: 0,
  }
}

function normalizeCostDimensions(value: unknown): CostDimensions {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) return {}
  const source = value as Record<string, unknown>
  const out: CostDimensions = {}
  for (const key of COST_DIMENSION_KEYS) {
    const v = source[key]
    if (typeof v === 'string' && v !== '') out[key] = v
  }
  return out
}

function numberValue(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined
}

function round8(value: number): number {
  return Math.round(value * 1e8) / 1e8
}
